lib/Runtime/ByteCode/ByteCodeEmitter.cpp

```diff
@@ -70,7 +70,7 @@ static void EmitOptionalChainWrapper(ParseNodeUni *pnodeOptChain, ByteCodeGenera
     // Acquire slot for the result value    
     Js::RegSlot resultSlot = funcInfo->AcquireLoc(pnodeOptChain);
     // Copy chain result
-    byteCodeGenerator->Writer()->Reg2(Js::OpCode::Ld_A_ReuseLoc, resultSlot, innerNode->location);
+    byteCodeGenerator->Writer()->Reg2(Js::OpCode::Ld_A, resultSlot, innerNode->location);
```

Comment:
Minor but I think you can delete this line.

resultSlot should be guaranteed to be the same register as innerNode->location here; I tested offline removing this line and your tests all still pass - except for the existing fail for the Simple method Calls test in the Jit.

Reply:
There's an edge case:
`ByteCodeGenerator::ReturnRegister == resultSlot` => `innerNode->location != resultSlot`

Reply:
Ugh didn't think of that. I suppose for optimal could do: `if (innerNode->location != resultSlot)`

Reply:
I'd just go back to sth like
```c++
innerNode->location = funcInfo->AcquireLoc(pnodeOptChain);
```

Reply:
Question: Is it bad to reserve a register BEFORE emitting the whole chain-expression?

Reply:
Ah, yes putting that before `emitChainContent(innerNode);` should dodge the issue.

Reply:
But this register can no longer be used by other sub-expressions.
Normally a register would be reserved after everything is done...

Reply:
> Question: Is it bad to reserve a register BEFORE emitting the whole chain-expression?

If you do:
```cc
funcInfo->AcquireLoc(someNode);
funcInfo->AcquireLoc(someNode);
funcInfo->AcquireLoc(someNode);
funcInfo->AcquireLoc(someNode);
```
The subsequent calls are going to return the same value as the first call - as it's cached at someNode->location and AcquireLoc checks that value and returns it if it's already been set.

Similarly if you do:
```cc
someNode->location = funcInfo->AcquireLoc(someOtherNode);
funcInfo->AcquireLoc(someNode); // returns existing value of someNode->location from previous line
```

Reply:
If there is nothing special about the node and it's not already got a Location assigned its location will just be set to the next available Temp, when you releaseLoc on a temp, that becomes the next available temp again hence why in the majority of cases your previous code was having the same Reg for innerNode and resultLoc.

Reply:
I know, but the `funcInfo->AcquireLoc(innerNode)` call would normally follow after the emission of the whole chain, now it's done before.

Normally the chain should be able to use `R7` but now it can't because `R7` was already reserved
```asm
    0046   ProfiledLdFld        R8 = R8.a #4 <4>
    004a   ProfiledLdFld        R8 = R8.b #5 <5>
    004e   ProfiledLdFld        R8 = R8.c #6 <6>
    0052   ProfiledLdFld        R8 = R8.d #7 <7>
    0056   ProfiledLdFld        R8 = R8.e #8 <8>
    005a   ProfiledLdFld        R8 = R8.f #9 <9>
    005e   ProfiledLdFld        R8 = R8.g #10 <10>
    0062   ProfiledLdFld        R7 = R8.h #11 <11>
```

Reply:
I thought in the version before you did `funcInfo->ReleaseLoc(innerNode)` but didn't actually do AcquireLoc for it as it happened inside the emitting of the chain.

I'm sorry this has brought in quite a bit of confusion, and I'm sorry I didn't think of the case where the resultSlot was already allocated (e.g. as the return reg) - I think it may be cleanest to return to the logic before but simply include an `if(resultSlot != innerNode->location)` around the LdA.

Reply:
Ooh, no need to say sorry; I just did not understand 😅
After some testing I actually think we need
```c++
innerNode->location = funcInfo->AcquireLoc(pnodeOptChain);
```
though, because the stack of temp-registers will be asymmetrical released for optional-chains as call-target (another edge-case)
```js
(a?.b)().c
```

Reply:
During testing, test262 sometimes failed while my own tests passed -> I need more tests!
(See ToDo in PR description)

Reply:
Thinking of extra tests, do you have a cases for an optional call where the property exists but is not callable?

Reply:
Now I have 😉